_posts/2016-06-22-beyond-progressive-web-apps-part-3.md

```diff
@@ -0,0 +1,229 @@
+But now imagine both author and editor fixed the typo, but the author decided to use a different word altogether, while the editor just corrected the spelling, and both are doing this at the same time.
+
+In that case, it is impossible for a computer to know which is the *correct* version. We could apply a policy in *our* app that in case of a conflict the author’s changes should take precedent. Or editors get the preferred treatment. Whichever it is, this can be a viable strategy for specific apps.
+
+As mentioned before, these are all examples, and we are trying to become experts in sync more generically. So we need a better solution.
+
+And here is a bit of a bummer: there are some types of conflicts that no computer can ever solve. We will need human intervention. There is no way around this, but at the end of the day, our brains are a lot more sophisticated than computers and *conflict resolution* is where this plays out.
+
+Despite this, I hope you feel a little more comfortable with conflicts.
+
+As an added bonus, now we are equipped to learn how to improve on timestamps for versioning our objects.
+
+## Vector Clocks
+
+If you start searching for techniques to solve the issues with timestamps from a wall clock, you will eventually find a mention of [vector clocks](https://en.wikipedia.org/wiki/Vector_clock), for a variant, see [Lampert Timestamps](https://en.wikipedia.org/wiki/Lamport_timestamps). A vector clock also works with timestamps, but its source is not a wall clock, like we humans use, but a so called *logical clock*.
+
+When using vector clocks as version specifiers, we not only send logical timestamps with our objects, but also the state of the logical clock, so that later, when it is *time* (pun intended), to compare two objects, we can calculate which one came first. Great!
+
+Just one more thing.
+
+![Lt. Frank Columbo, LAPD, Homicide](/dist1/blog/2016/06/columbo.jpg)
+
+Remember when we talked about the author and the editor of a story fixing a typo and adding to the story at the same time? We now know to detect and how to resolve this conflict.
+
+But what if the typo fix both create result in the *same* text? With timestamps or vector clocks, a conflict is still generated, and our subsequent merge procedure has got nothing to do. So we can handle this too, but since the resulting story is the same, wouldn’t it be nice to *not* create a conflict here in the first place?
+
+I realise this example is a little bit contrived, but I hope by staying within our example project, we can see the problem here. Where the problem would *actually* make our lives harder is when we don’t have a single sync server, but a cluster of sync servers, which we might need for scaling to higher and higher loads of our system.
+
+Explaining all this takes time, and this tutorial is already getting long, so we are going to skip the nitty-gritty details of that one. I hope you forgive me. If you don’t, I hope this gif of a squirrel makes up for it:
```

Comment:
only saw this after hitting the merge button, fixed on main branch, thanks for the heads up!